working/1426-extension-types/feature-specification.md

```diff
@@ -0,0 +1,884 @@
+restrict the access to that representation to a set of operations that are
+independent of the operations available on the representation. In other
+words, the extension type ensures that we only work with the representation
+in specific ways, even though the representation itself has an interface
+that allows us to do many other (wrong) things.
+
+It would be straightforward to enforce an added discipline like this by
+writing a wrapper class with the allowed operations as members, and
+working on a wrapper object rather than accessing `o` and its methods
+directly:
+
+```dart
+// Attempt to emulate the extension type using a class.
+
+class TinyJson {
+  // `representation` is assumed to be a nested list of numbers.
+  final Object representation;
+
+  TinyJson(this.representation);
+
+  Iterable<num> get leaves sync* {
+    var self = representation;
+    if (self is num) {
+      yield self;
+    } else if (self is List<dynamic>) {
+      for (Object element in self) {
+        yield* TinyJson(element).leaves;
+      }
+    } else {
+      throw "Unexpected object encountered in TinyJson value";
+    }
+  }
+}
+
+void main() {
+  TinyJson tiny = TinyJson(<dynamic>[<dynamic>[1, 2], 3, <dynamic>[]]);
+  print(tiny.leaves);
+  tiny.add("Hello!"); // Error.
+}
+```
+
+This is similar to the extension type in that it enforces the use of
+specific operations (here we only have one: `leaves`) and in general makes
+it an error to use instance methods of the representation (e.g., `add`).
+
+Creation of wrapper objects takes time and space, and in the case where we
+wish to work on an entire data structure we'd need to wrap each object as
+we navigate the data structure. For instance, we need to create a wrapper
+`TinyJson(element)` in order to invoke `leaves` recursively.
+
+In contrast, the extension type mechanism is zero-cost, in the sense that
+it does _not_ use a wrapper object, it enforces the desired discipline
+statically.
+
+Like extension methods, extension types are static in nature: An extension
+type declaration may declare some type parameters (just like the current
+extension declarations). The type parameters will be bound to a type which
+is determined by the static type of the receiver. Similarly, like extension
+methods, members of an extension type are resolved statically, i.e., if
+`tiny.leaves` is an invocation of an extension type getter `leaves`, then
+the declaration named `leaves` whose body is executed is determined at
+compile-time. There is no support for late binding of an extension method,
+and hence there is no notion of overriding. In return for this lack of
+expressive power, we get improved performance.
+
+Here is another example. It illustrates the fact that an extension type `E`
+with on-type `T` introduces a type `E` which is a supertype of `T`. This
+makes it possible to assign an expression of type `T` to a variable of type
+`E`. This corresponds to "entering" the extension type (accepting the
```

Comment:
I'm really very skeptical that this should be the default, assuming it's allowed.  It feels very error prone to not get any type discipline to help me.  With your TinyJson example, any assignment that I write to `tiny` will be valid (I guess as long as it's a non-nullable type).  That doesn't feel like what I want in the static analysis by default (or possibly at all).

Reply:
> The rationale would be that if an object `o` is accessed with static type `E` then it is guaranteed to be an instance of the associated on-type `T`, and it is always safe to invoke `E` members with such an object as `this`.

Sure.  And it's always safe to invoke `T` members on an `E`. Safety isn't the issue - maintaining abstraction is.  



> I tend to think that developers will need to make the trade-off: Do we use a protected extension type (and spend the type running a constructor)

By "type" you mean "time"?  There need be no time involved.  If the constructor is trivial it is zero cost - it's a statically dispatched empty bodied function.  If the compiler doesn't inline that away that we have serious problems.

Reply:
> Safety isn't the issue - maintaining abstraction is.

But that's exactly the point: For an object _o_ whose run-time type `T0` is a subtype of the on-type `T` of an extension type `E`, we maintain a specific discipline by accessing _o_ with static type `E` (above and beyond the discipline imposed by accessing it with static type `T0` or `T`), because everything we can with _o_ do is expressed in terms of the interface of `T`, and we can only do those things that the member implementations of `E` choose to do using that interface.

So a cast from `T` to `E` imposes the abstraction, and it can occur implicitly. A cast from `E` to `T` breaks the abstraction, and it must be explicit.

I would describe that as 'maintaining the abstraction' (even though the required explicit cast is of course a very weak kind of protection).

> By "type" you mean "time"?

Yes, I meant "Do we use a protected extension type (and spend the time running a constructor) .."

> There need be no time involved. If the constructor
> is trivial it is zero cost - it's a statically dispatched
> empty bodied function.

Agreed, I'd also hope that we can consistently optimize many cases like this.

I mentioned the time taken to run a non-trivial constructor because I believe that's the real alternative: A protected extension type would be used in the situation where we are willing to pay (in terms of resource consumption at run time) for invariant verification that requires actual computations.

If we do not wish to maintain or check any non-trivial invariants (so every instance of the on-type is good enough to be `this` for every member of the extension) then it's unlikely that we would use a protected extension type.

It's of course also possible that we'd want computational invariant verification during development, and then we'd use informal program proofs together with the run-time checks to support the assumption that we won't actually break those invariants, and then we'd use a zero-cost constructor in production code.

That's the trade-off I was referring to.